model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/scaled_dot_product_attention.py

```diff
@@ -0,0 +1,187 @@
+import torch.nn as nn
+import torch
+import math
+from copy import copy
+import numpy as np
+from model_compression_toolkit.core.common.graph.functional_node import FunctionalNode
+from model_compression_toolkit.core.common import BaseSubstitution
+from model_compression_toolkit.core.common.graph.graph_matchers import NodeOperationMatcher
+from model_compression_toolkit.core.common.graph.base_graph import Graph, BaseNode, OutTensor
+from model_compression_toolkit.core.pytorch.constants import DIM
+from model_compression_toolkit.core.pytorch.pytorch_device_config import get_working_device
+
+
+class ScaledDotProductDecomposition(BaseSubstitution):
+    """
+    Removes a scaled_dot_product_attention node from the graph, and replaces it with a compatible graph that
+    consists of:
+    MatMul(over q and transposed k)
+    Mul (for scaling)
+    Add (for masking. optional operation, used in cases that attn_mask ig given)
+    Dropout
+    Softmax
+    Matmul.
+    """
+
+    def __init__(self):
+        """
+        Matches scaled_dot_product_attention node.
+        """
+        super().__init__(matcher_instance=NodeOperationMatcher(nn.functional.scaled_dot_product_attention))
+
+    def _get_input_by_name(self, attention_node, input_name, input_index, default_value):
+        """
+        Search for attention_node input value in op_call_kwargs (using input_name) and op_call_args (using input_index).
+        In case the input is not given, returns its default_value.
+
+        """
+        if input_name in attention_node.op_call_kwargs:
+            return attention_node.op_call_kwargs[input_name]
+        elif len(attention_node.op_call_args) > input_index:  # input order: [attn_mask, dropout_p, is_causal]
+            return attention_node.op_call_args[input_index]
+        return default_value
+
+    def _get_attention_input_nodes(self, graph: Graph, attention_node: BaseNode) -> dict:
+        q, k, v = 0, 1, 2
+        prev_nodes = graph.get_prev_nodes(attention_node)
+        q_node, k_node, v_node = prev_nodes[q], prev_nodes[k], prev_nodes[v]
+        assert q_node.name == "q" and k_node.name == "k" and v_node.name == "v", \
```

Comment:
We are not allowing any over names for the input nodes?

Reply:
you are correct, I removed the assertion 👍